### PR TITLE
Automatically push Docker image to Docker Hub

### PR DESCRIPTION
**Pull request by ErikSchierboom**
The [v3 development environment](https://github.com/exercism/development-environment/) uses Docker images from Docker Hub for all the tooling.
This allows for easy testing of the various tooling components (analyzers/representer/test-runners).

This PR adds a GitHub Actions workflow to automatically build and push a new Docker image whenever something is merged to master.

See https://github.com/exercism/v3/issues/2727

### CODE REVIEW

**Review by ynfle — APPROVED** [review]
I'm assuming that this works. So, 👍

File: .github/workflows/dockerhub.yml [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+jobs:
+  dockerhub:
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v2
+
+      - uses: satackey/action-docker-layer-caching@v0.0.5
```
Comment (ee7) [inline .github/workflows/dockerhub.yml:15]:
```suggestion
      - uses: satackey/action-docker-layer-caching@v0.0.8
```

This suggestion updates it to the latest version, but should we pin this to some hash instead? I see that this is a relatively obscure action. What's our general policy on pinning at the moment? Only pin when there are secrets involved?

File: .github/workflows/dockerhub.yml [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+      - name: Build and push Docker images
+        uses: docker/build-push-action@v1.1.0
+        with:
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_PASSWORD }}
+          repository: ${{ github.event.repository.full_name }}
+          tag_with_sha: true
+          tags: latest
+          dockerfile: Dockerfile
```
Comment (ee7) [inline .github/workflows/dockerhub.yml:25]: Looks like we should upgrade to v2, but we can do that in a later PR - it seems like a big release that requires significant changes. See the [upgrade notes](https://github.com/docker/build-push-action/blob/master/UPGRADE.md#simple-workflow).

In particular, the tagging part seems much more involved in v2.

Don't accept the below suggestion - it won't work. But I think it's close.

```suggestion
      - name: Docker meta
        id: docker_meta
        uses: crazy-max/ghaction-docker-meta@v1
        with:
          images: foo/bar # ?
          
      - name: Set up Docker Buildx
        uses: docker/setup-buildx-action@v1

      - name: Login to DockerHub
        uses: docker/login-action@v1
        with:
          username: ${{ secrets.DOCKER_USERNAME }}
          password: ${{ secrets.DOCKER_PASSWORD }}

      - name: Build and push Docker images      
        uses: docker/build-push-action@v2.1.0
        with:
          file: ./Dockerfile
          tags: ${{ steps.docker_meta.outputs.tags }}
          labels: ${{ steps.docker_meta.outputs.labels }}
```

**Review by ee7 — COMMENTED** [review]
I'm fine with merging this - I understand that it's known to work.

But I have some questions about version pinning, and did anybody look into upgrading the `docker/build-push-action` to v2?

File: .github/workflows/dockerhub.yml (hunk 2 above)
Comment (ErikSchierboom) [reply .github/workflows/dockerhub.yml:25]: Interesting. I didn't know they changed it.

File: .github/workflows/dockerhub.yml (hunk 1 above)
Comment (ErikSchierboom) [reply .github/workflows/dockerhub.yml:15]: We should definitely pin. This is an oversight on my part.

Comment (ee7) [inline .github/workflows/dockerhub.yml:15]:
```suggestion
      - uses: satackey/action-docker-layer-caching@da67a6cd88114ce3e232893dc6d946efa225aefb # v0.0.8
```

Comment (ee7) [reply .github/workflows/dockerhub.yml:15]: Okay - I'll update my suggested change then.

File: .github/workflows/dockerhub.yml [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v2
+
+      - uses: satackey/action-docker-layer-caching@v0.0.5
+
+      - name: Build and push Docker images
+        uses: docker/build-push-action@v1.1.0
```
Comment (ee7) [inline .github/workflows/dockerhub.yml:18]:
```suggestion
        uses: docker/build-push-action@3e7a4f6646880c6f63758d73ac32392d323eaf8f # v1.1.2
```

File: .github/workflows/dockerhub.yml (hunk 2 above)
Comment (ee7) [reply .github/workflows/dockerhub.yml:25]: v2 was only released 2 weeks ago. In the latest v1 they've added a deprecation notice that advises upgrading to v2, but that wasn't included in the v1.1.0 we were using. I'll make a suggestion that just pins to the v1.1.2 hash for this PR.

**Review by ee7 — COMMENTED** [review]
Pinning to hashes.

**Comment by iHiD** [comment]
(Someone (one of you?) is trying the nim test runner on staging. It won't work until this is merged FYI.)

**Comment by ee7** [comment]
> (Someone (one of you?) is trying the nim test runner on staging. It won't work until this is merged FYI.)

Not me. But I'll merge this PR.

**Review by iHiD — APPROVED** [review]

**Comment by ynfle** [comment]
Was me. Forgot about this. I apologize for the headache, I know you've got ~ 1 million things to deal with.

Thanks for following up

**Pull request merged**